### ISSUE TITLE
how to debug the casadi optimisation

### ISSUE THREAD

**Issue by woodwm** [FilippoAiraldi/metanet#2]
Hi,

I notice that in the first two optimisations in the script `coordinated_control_in_casadi.py`, it fails to get a good result as it reaches the maximum iteration number. In this case, how to find out which constraints are violated? Even `casadi.opti` has few debug functions, it is still challenging to find out why the optimisation fails as I do not know how to output the internal values during the optimisation.

I changed to metanet parameters to my calibrated parameters. It happens even more often that the optimisation reaches the maximum iteration number. I wonder what can be done in this case. 

Increasing the maximum iteration number does not help mostly. I highly suspect it is because metanet outputs infeasible result.

**Comment by FilippoAiraldi** [comment]
Hello,

I just run again the script on my laptop, and was unable to reproduce your problem. However, I have seen that different machines, e.g., depending on the OS (Windows or Linux) or CasADi version, tend to lead to different results of the MPC optimization.

I am not an expert, but usually when IPOPT reaches the maximum number of iterations, you have to take a look and understand what is preventing the solver from converging. For example, take a look at [this](https://list.coin-or.org/pipermail/ipopt/2007-February/000699.html) if the dual variables are the culript. In general, the larger and more nonlinear the problem gets, the more likely these errors are and the more difficult it is to debug.

In practice, based on my experience, I can only suggest two things
1. **IPOPT**: playing around with its [options](https://coin-or.github.io/Ipopt/OPTIONS.html) might help. Not only the maximum number of iterations, but also the various tolerances (especially if you don't need very high accuracy in your primal-dual solution), etc.
2. **METANET**: its dynamics are highly nonlinear; moreover, some states like the average speeds tend to become negative, which is obviously unrealistic. For this reason, I strongly suggest to play with the optional keywords of the [`Network.step`](https://github.com/FilippoAiraldi/sym-metanet/blob/d3da906cf3ca1376714761d2fff2677278a991bd/src/sym_metanet/network.py#L508) method to force some or all of these quantities to positive. While this operation adds an extra nonlinearity (in the form of an additional $max{(\cdot, 0)}$), I've seen it can greatly help the IPOPT solver in the MPC scheme

**Comment by woodwm** [comment]
I run again the script `coordinated_control_in_casadi.py` on windows 11 and casadi 3.6.3. It only reaches the maximum iteration number when `k=0`. Probably you are right this is because of machine problem.

After changing to my metanet parameters, this happens more often. When it reaches the maximum iteration number, I see that it is because the primal problem does not converge and the constraint violation is high. 

I definitely agree that the main trouble is that link speed may becomes negative under some parameters and traffic state.

I accatually used your old version of this package before it is reformed. I used the OPTI stack, and I found that when it is unable to find the result, it is mainly because `opti.subject_to(vars[f'v_{link}'][:, k + 1] == outs[i + 1])` are voilated
```shell
------- i = 3049/4535 ------ 
0 <= 2.61051 <= 0 (viol 2.61051)
Opti constraint of shape 4x1, defined at \metanet\ctrl\solvers.py:889 in __create_constraints
  opti.subject_to(vars[f'v_{link}'][:, k + 1] == outs[i + 1])
At nonzero 2.
```
`F = sim2func(sim, out_nonneg=True)` is forced to be nonnegtive. I tried to  get the system state after IPOPT reaches the maximum iteration number by using its outputed control result in the last iteration, I couldn't find the case when any traffic state variable (density and speed) becomes zero. That is very strange.

Thanks a lot for your advice.
